sdk/tests/storage/blobs/test_az_storage_blobs.c

```diff
@@ -312,3 +332,258 @@ void test_storage_blobs_download(void** state)
+  memset(url_buf + URL_START_LEN, 'y', sizeof(url_buf) - URL_START_LEN);
+
+  assert_int_equal(
+      AZ_ERROR_NOT_ENOUGH_SPACE,
+      az_storage_blobs_blob_client_init(
+          &client, AZ_SPAN_FROM_BUFFER(url_buf), AZ_CREDENTIAL_ANONYMOUS, NULL));
+}
+
+#undef URL_START
+#undef URL_START_LEN
+
+static AZ_NODISCARD az_result test_credential_fn(void* ref_credential, az_span scopes)
+{
+  (void)ref_credential;
+  (void)scopes;
+
+  return AZ_ERROR_UNEXPECTED_CHAR;
+}
+
+void test_storage_blobs_init_credential_error(void** state);
+void test_storage_blobs_init_credential_error(void** state)
+{
+  (void)state;
+
+  _az_credential cred = (_az_credential){ ._internal = { .set_scopes = test_credential_fn } };
+
+  az_storage_blobs_blob_client client = { 0 };
+  assert_int_equal(
+      AZ_ERROR_UNEXPECTED_CHAR,
+      az_storage_blobs_blob_client_init(&client, AZ_SPAN_FROM_STR("x://y"), &cred, NULL));
+}
+
+static az_result no_op_transport(az_http_request const* request, az_http_response* ref_response)
+{
+  (void)request;
+
+  return az_http_response_init(
+      ref_response,
+      AZ_SPAN_FROM_STR("HTTP/1.1 200 OK\r\n"
+                       "Content-Length: 0\r\n"
+                       "\r\n"));
+}
+
+void verify_storage_blobs_upload_empty_host(void** state);
+void verify_storage_blobs_upload_empty_host(void** state)
+{
+  (void)state;
+
+  az_storage_blobs_blob_client client = { 0 };
+  assert_true(az_result_succeeded(az_storage_blobs_blob_client_init(
+      &client, AZ_SPAN_FROM_STR("x:///"), AZ_CREDENTIAL_ANONYMOUS, NULL)));
+
+  uint8_t response_buffer[1024 * 4] = { 0 };
+  az_http_response response = { 0 };
+  assert_true(
+      az_result_succeeded(az_http_response_init(&response, AZ_SPAN_FROM_BUFFER(response_buffer))));
+
+  _az_http_client_set_callback(no_op_transport);
+
+  assert_true(az_result_succeeded(az_storage_blobs_blob_upload(
+      &client, NULL, AZ_SPAN_FROM_STR("BlobContent"), NULL, &response)));
+
+  _az_http_client_set_callback(NULL);
+}
+
+#ifndef AZ_NO_PRECONDITION_CHECKING
+
+ENABLE_PRECONDITION_CHECK_TESTS()
+
+void test_storage_blobs_init_null_client(void** state);
+void test_storage_blobs_init_null_client(void** state)
+{
+  (void)state;
+  SETUP_PRECONDITION_CHECK_TESTS();
+
+  ASSERT_PRECONDITION_CHECKED(
+      az_storage_blobs_blob_client_init(NULL, AZ_SPAN_EMPTY, AZ_CREDENTIAL_ANONYMOUS, NULL));
+}
+
+void test_storage_blobs_init_bad_url(void** state);
+void test_storage_blobs_init_bad_url(void** state)
+{
+  (void)state;
+  SETUP_PRECONDITION_CHECK_TESTS();
+
+  az_storage_blobs_blob_client client = { 0 };
+  ASSERT_PRECONDITION_CHECKED(
+      az_storage_blobs_blob_client_init(&client, AZ_SPAN_EMPTY, AZ_CREDENTIAL_ANONYMOUS, NULL));
+}
+
+void verify_storage_blobs_upload_null_client(void** state);
+void verify_storage_blobs_upload_null_client(void** state)
+{
+  (void)state;
+  SETUP_PRECONDITION_CHECK_TESTS();
+
+  _az_http_client_set_callback(no_op_transport);
```

Comment:
Unrelated question: Is this how we expect customers to set the transport?

Reply:
To pick curl, you specify `TRANSPORT_CURL=ON` when building the SDK. There isn't any other transport at the moment, but in general the transport should implement the  `az_http_client_send_request()` function. If there was WinHTTP implementation, for example, I'd think there would be some switch like `TRANSPORT_WINHTTP=ON`, and then in CMakeLists.txt for core we would replace `az_curl.c` with `az_winhttp.c` (can't easily have both, although it is possible).

In this specific case, when storage tests are built, they do implement the `az_http_client_send_request()`, and `_az_http_client_set_callback()` is really a part of storage tests, not the core API: see https://github.com/Azure/azure-sdk-for-c/blob/feature/storage-blobs/sdk/tests/storage/blobs/_az_test_http_client.h and https://github.com/Azure/azure-sdk-for-c/blob/feature/storage-blobs/sdk/tests/storage/blobs/_az_test_http_client.c . It allows you to provide callback, but our core does not directly offer you that feature, although as you see, if test code can do that, core could add that in the future if needed, or even customers can add that feature themselves if they need to switch between transport (although they'll have to do some linker magic or code patching to rename az_curl's `az_http_client_send_request()`.